libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1263,7 +1265,41 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 				// as require with custom errors is not supported in legacy codegen.
 				auto const* magicType = dynamic_cast<MagicType const*>(arguments[1]->annotation().type);
 				if (magicType && magicType->kind() == MagicType::Kind::Error)
-					solUnimplemented("Require with a custom error is only available using the via-ir pipeline.");
+				{
+					// Make sure that error constructor arguments are evaluated regardless of the require condition
+					auto const& errorConstructorCall = dynamic_cast<FunctionCall const&>(*arguments[1]);
+					errorConstructorCall.expression().accept(*this);
+					std::vector<Type const*> errorConstructorArgumentTypes{};
+					unsigned stackDepth = 0;
+					for (ASTPointer<Expression const> const& errorConstructorArgument: errorConstructorCall.sortedArguments())
+					{
+						errorConstructorArgument->accept(*this);
+						stackDepth += errorConstructorArgument->annotation().type->sizeOnStack();
+						errorConstructorArgumentTypes.push_back(errorConstructorArgument->annotation().type);
+					}
+					// stack: <condition> <arg0> <arg1> ... <argN>
+					// Move condition to the top of the stack
+					utils().moveIntoStack(arguments.at(0)->annotation().type->sizeOnStack(), stackDepth);
```

Comment:
We have `moveToStackTop()`, which is just `moveIntoStack()` with swapped arguments and looks like it would make your code much more self-explanatory.

I'd also rename `stackDepth` to `sizeOfErrorArguments` because it explains its purpose much better.

And some other small tweaks:
```suggestion
					for (ASTPointer<Expression const> const& errorArgument: errorConstructorCall.sortedArguments())
					{
						errorArgument->accept(*this);
						errorArgumentTypes.push_back(errorArgument->annotation().type);
					}
					unsigned sizeOfCondition = arguments.at(0)->annotation().type->sizeOnStack();
					unsigned sizeOfErrorArguments = CompilerUtils::sizeOnStack(errorArgumentTypes);
					// stack: <condition> <arg0> <arg1> ... <argN>
					utils().moveToStackTop(sizeOfErrorArguments, sizeOfCondition);
					// stack: <arg0> <arg1> ... <argN> <condition>
```

Reply:
I see you're no longer passing in `arguments.at(0)->annotation().type->sizeOnStack()`, relying on the default value of the parameter I guess?

Which is fine, but in that case I'd still assert that it's equal to 1 to be safe.

Reply:
Or maybe it's fine as is... If this was violated probably many more things would go wrong here.